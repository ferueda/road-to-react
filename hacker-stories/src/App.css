button:hover > svg > g {
  fill: #fff;
  stroke: #fff;
}
