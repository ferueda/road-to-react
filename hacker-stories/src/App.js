import React, { useState } from 'react';

const ListItem = ({ title, url, author, num_comments, points }) => {
  return (
    <div>
      <span>
        <a href={url} target='_blank' rel='noopener noreferrer'>
          {title}
        </a>{' '}
        by {author}
      </span>
      <br />
      <span>Comments: {num_comments}</span>
      <br />
      <span>Points: {points}</span>
      <hr />
    </div>
  );
};

const List = ({ list }) => {
  return list.map(({ objectID, ...item }) => (
    <ListItem key={objectID} {...item} />
  ));
};

const Search = ({ search, onSearch }) => {
  return (
    <div>
      <label htmlFor='search'>Search: </label>
      <input id='search' type='text' value={search} onChange={onSearch} />
      <p>
        Searching for <strong>{search}</strong>
      </p>
    </div>
  );
};

const App = () => {
  const [search, setSearch] = useState('');
  const handleSearch = (event) => {
    setSearch(event.target.value);
  };

  const stories = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];

  return (
    <div className='App'>
      <h1>My Hacker Stories</h1>
      <Search search={search} onSearch={handleSearch} />
      <hr />
      <List
        list={stories.filter((story) =>
          story.title.toLowerCase().includes(search.toLowerCase())
        )}
      />
    </div>
  );
};

export default App;
