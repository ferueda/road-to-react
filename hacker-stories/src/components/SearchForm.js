import React, { useRef, useEffect } from 'react';

const InputWithLabel = ({
  type = 'text',
  id,
  children,
  value,
  onInputChange,
  isFocused,
}) => {
  const inputRef = useRef();

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <React.Fragment>
      <label htmlFor={id}>{children}</label>
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </React.Fragment>
  );
};

const SearchForm = ({ handleSearchInput, search, handleSearchSubmit }) => {
  return (
    <form onSubmit={handleSearchSubmit}>
      <InputWithLabel
        type='text'
        value={search}
        id='search'
        isFocused
        onInputChange={handleSearchInput}
      >
        <strong>Search: </strong>
      </InputWithLabel>
      <button type='submit' disabled={!search}>
        Search
      </button>
    </form>
  );
};

export default React.memo(SearchForm);
