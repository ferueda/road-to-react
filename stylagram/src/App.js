import React from 'react';
import styled from 'styled-components';
import Header from './components/Header';
import Profile from './components/Profile';
import GlobalStyle from './theme/globalStyle';

const Button = styled.button`
  cursor: pointer;
  background: transparent;
  font-size: 16px;
  border-radius: 3px;
  color: palevioletred;
  color: ${({ primary }) => (primary ? 'violet' : 'palevioletred')};
  border: ${({ primary }) =>
    primary ? '2px solid violet' : '2px solid palevioletred'};
  margin: 0 1em;
  padding: 0.25em 1em;
  transition: 0.5s all ease-out;

  &:hover {
    color: white;
    background-color: ${({ primary }) =>
      primary ? 'violet' : 'palevioletred'};
  }
`;

const GreenButton = styled(Button)`
  background-color: white;
  color: green;
  border-color: green;
`;

const AppWrapper = styled.div`
  background-color: #fafafa;
`;

function App() {
  return (
    <React.Fragment>
      <GlobalStyle />
      <AppWrapper>
        <Header />
        <Profile />
      </AppWrapper>
    </React.Fragment>
  );
}

export default App;
