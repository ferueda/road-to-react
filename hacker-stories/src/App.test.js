import React from 'react';
import renderer from 'react-test-renderer';
import axios from 'axios';

jest.mock('axios');

import App from './App';
import MemoizedList, { ListItem as MemoizedListItem } from './components/List';
import SearchForm, { InputWithLabel } from './components/SearchForm';

const ListItem = MemoizedListItem.type;
const List = MemoizedList.type;

describe('ListItem', () => {
  const item = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };

  const handleRemoveItem = jest.fn();
  let component;

  beforeEach(() => {
    component = renderer.create(
      <ListItem item={item} onRemoveItem={handleRemoveItem} />
    );
  });

  it('renders all properties', () => {
    expect(component.root.findByType('a').props.href).toEqual(
      'https://reactjs.org/'
    );
  });

  it('calls onRemoveItem on button click', () => {
    component.root.findByType('button').props.onClick();
    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    expect(handleRemoveItem).toHaveBeenCalledWith(item);
  });

  test('renders snapshot', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('List', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];

  const handleRemoveItem = jest.fn();
  let component;

  beforeEach(() => {
    component = renderer.create(
      <List list={list} onRemoveItem={handleRemoveItem} />
    );
  });

  it('renders two items', () => {
    expect(component.root.findAllByType(ListItem).length).toEqual(2);
  });

  it('it calls onRemoveItem in every item', () => {
    component.root.findAllByType('button').forEach((b) => b.props.onClick());
    expect(handleRemoveItem).toHaveBeenCalledTimes(2);
  });
});

describe('SearchForm', () => {
  const searchFormProps = {
    search: 'React',
    handleSearchInput: jest.fn(),
    handleSearchSubmit: jest.fn(),
  };

  let component;

  beforeEach(() => {
    component = renderer.create(<SearchForm {...searchFormProps} />);
  });

  it('renders the input field with its value', () => {
    const value = component.root.findByType(InputWithLabel).props.value;
    expect(value).toEqual('React');
  });

  it('changed the input field', () => {
    const pseudoEvent = { target: 'Redux' };

    component.root.findByType('input').props.onChange(pseudoEvent);

    expect(searchFormProps.handleSearchInput).toHaveBeenCalledTimes(1);
    expect(searchFormProps.handleSearchInput).toHaveBeenCalledWith(pseudoEvent);
  });

  it('submits the form', () => {
    const pseudoEvent = {};

    component.root.findByType('form').props.onSubmit(pseudoEvent);

    expect(searchFormProps.handleSearchSubmit).toHaveBeenCalledTimes(1);
    expect(searchFormProps.handleSearchSubmit).toHaveBeenCalledWith(
      pseudoEvent
    );
  });

  it('disables the button and prevents submit', () => {
    component.update(<SearchForm {...searchFormProps} search='' />);

    expect(component.root.findByType('button').props.disabled).toBeTruthy();
  });
});

describe('App', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];

  let component;

  it('succeeds fetching data with a list', async () => {
    const promise = Promise.resolve({
      data: {
        hits: list,
      },
    });

    axios.get.mockImplementationOnce(() => promise);

    await renderer.act(async () => {
      component = renderer.create(<App />);
    });

    expect(component.root.findByType(List).props.list).toEqual(list);
  });

  it('fails fetching data with a list', async () => {
    const promise = Promise.reject();

    axios.get.mockImplementationOnce(() => promise);

    await renderer.act(async () => {
      component = renderer.create(<App />);
    });

    expect(component.root.findByType('p').props.children).toEqual(
      'Something went wrong...'
    );
  });
});
