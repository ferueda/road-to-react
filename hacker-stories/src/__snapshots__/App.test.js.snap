// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders snapshot 1`] = `
<div>
  <span>
    <a
      href="https://reactjs.org/"
      rel="noopener noreferrer"
      target="_blank"
    >
      React
    </a>
     
    by 
    Jordan Walke
  </span>
  <br />
  <span>
    Comments: 
    3
  </span>
  <br />
  <span>
    Points: 
    4
  </span>
  <button
    onClick={[Function]}
    type="button"
  >
    Dismiss 
    <svg
      height="14px"
      width="14px"
    >
      check.svg
    </svg>
  </button>
  <hr />
</div>
`;
